aptos-move/aptos-vm-types/src/change_set.rs

```diff
@@ -345,13 +388,55 @@ impl VMChangeSet {
         Ok(())
     }
 
+    fn squash_group_writes(
+        write_set: &mut HashMap<StateKey, GroupWrite>,
+        additional_write_set: HashMap<StateKey, GroupWrite>,
+    ) -> anyhow::Result<(), VMStatus> {
+        for (key, additional_update) in additional_write_set.into_iter() {
+            match write_set.entry(key) {
+                Occupied(mut group_entry) => {
+                    let GroupWrite {
+                        metadata_op: additional_metadata_op,
+                        inner_ops: additional_inner_ops,
+                    } = additional_update;
+
+                    // Squashing creation and deletion is a no-op. In that case, we have to
+                    // remove the old GroupWrite from the group write set.
+                    let noop = !WriteOp::squash(
+                        &mut group_entry.get_mut().metadata_op,
+                        additional_metadata_op,
+                    )
+                    .map_err(|e| {
+                        VMStatus::error(
+                            StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                            err_msg(format!("Error while squashing two write ops: {}.", e)),
+                        )
+                    })?;
+                    if noop {
+                        group_entry.remove();
+                    } else {
+                        Self::squash_additional_writes(
```

Comment:
If metadata op is delete, they should all be delete (the summation logic in the next PR should ensure this invariant), so it's maybe useful to assert / check this at places.
If they are all delete the metadata op may not be a delete because there might be more untouched resources in the group.
the inner ops can be anything (create/modify/delete), it's basically whatever happens to the resource (it's just within a group)

Reply:
added a comment and will enforce this invariant when creating the metadata_op: if deletion, all inner ops should be deletions, if creation, there can't be a deletion.